### PR TITLE
Reset Password 1: Forget Password API

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
## Description

Enable users to reset their password.

This PR completes **1.** 
--
~~0. Add `reset_token_digest` string attribute and `reset_token_expires_at` to users model with no restrictions.~~
~~1. Make an API for the forget password that receives the user email:
	- Finds the user by email.
	- generate a secure random **UNIQUE** complex token.
	- saves the token digest in the user `reset_token_digest` and set the `reset_token_expires_at`.
	>NOTE: The `reset_token_expires_at `= Time.now + PERIOD(seconds) where PERIOD > 0
	, PERIOD by default will be 15mins in seconds.
	- Sends the user the reset email with the token  _[Upcoming milestones]_.~~

>Note: For security reasons only the digest will be saved in the database and the token will be created and sent on the run.

2. Make an API for the reset password that receives the token , the new password and the user email:
	- Takes the token and validate its authenticity and integrity.
	- Validate the token expiration.
	- Update the user password.
	- Remove the user reset token attributes _[If the user update password succeeds]_.
3. Create the forget password UI and integrate it with the forget password API.
4. Create the reset password UI and integrate it with the reset password API.

### User story [Reset password]:
---
1. User navigates to sign-in page.
2. User clicks the forget password.
3. User enters the forget password UI to be prompted for their email:

>    Scenarios:
> 	1. if the user email is correct then a reset email will be sent to the user while
> 	indicating to check their inbox or spam folder.
> 	2. if the user email is not found then no email will be sent to the user but the user
> 	**WILL STILL** be indicated as if the email was sent _[Obscurity]_.

4. User receives the link with the **token digest**.
5. User _[if accesses the link before expiration]_ will be prompted for a new password and its 
confirmation.
  >Constraints: No reset password API request will initiate until:
	1. All fields are filled.
	2. Password confirmation matches the new password.
	_Note: If GLv3 include a password policy then the new password should conform to it._
6. User will get redirected to login page with the proper feedbacks and/or notifications.
[OR]
6. User will get redirected to login page if something went wrong and authenticate 
with the new password on success.

## Testing Steps

> 1. Pull the code.
> 2. Install the dependencies `bundle install && npm|yarn install`.
> 3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
> 4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:drop && rails db:create && rails db:migrate:with_data`
> 5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec && npx eslint app/javascript/* --ext .jsx,.js`
> 6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

### CODE REVIEW

**Comment by KH-Amir-TN** [comment]
# DEPENDS ON #3531

File: app/models/user.rb [hunk 1]
```diff
@@ -30,6 +30,25 @@ def self.search(input)
     all
   end
 
+  # Generates a token digest.
+  def self.generate_digest(token)
+    Digest::SHA2.hexdigest(token)
+  end
```
Comment (farhatahmad) [inline app/models/user.rb:36]: For now this is fine, though we may want to move this to a service in the future

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -42,4 +42,30 @@
...
+      end
+    end
+  end
+
+  context 'static methods' do
+    describe '#generate_digest' do
+      it 'calls Digest::SHA2#hexdigest to generate a digest' do
+        allow(Digest::SHA2).to receive(:hexdigest).and_return(true)
+        expect(Digest::SHA2).to receive(:hexdigest).with('test')
+        described_class.generate_digest('test')
```
Comment (farhatahmad) [inline spec/models/user_spec.rb:67]: I would do this a different way 

```
expect(described_class.generate_digest('test')).to eq(Digest::SHA2.hexdigest('test'))

```

File: app/models/user.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:36]: Yup, we may generate more secure tokens and crypto sign there digest

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -88,6 +107,10 @@ def user_params
       def change_password_params
         params.require(:user).permit(:old_password, :new_password)
       end
+
+      def forget_password_params
+        params.require(:user).permit(:email)
+      end
```
[the remark this answers is not in the document]
Comment (KH-Amir-TN) [reply app/controllers/api/v1/users_controller.rb:113]: Agreed,

**Review by farhatahmad — APPROVED** [review]

**Pull request merged**